src/intellisense/texDocSymbolProvider.ts

```diff
@@ -0,0 +1,296 @@
+            case 'verbatim':
+            case 'mathenv':
+                return envToStr(node);
+            case 'inlinemath':
+                return `$${argContentToStr(node.content)}$`;
+            case 'displaymath':
+                return `\\[${argContentToStr(node.content)}\\]`;
+            case 'group':
+                return preserveCurlyBrace ? `{${argContentToStr(node.content)}}` : argContentToStr(node.content);
+            case 'verb':
+                return node.content;
+            default:
+                return '';
+        }
+    }).join('');
+}
+
+// reference: https://github.com/James-Yu/LaTeX-Workshop/blob/master/src/outline/structurelib/latex.ts#L30
+export async function construct(document: vscode.TextDocument): Promise<TeXElement[]> {
+    const filePath = document.fileName;
+    if (filePath === undefined) {
+        return [];
+    }
+    const structs: FileStructureCache = {};
+    await constructFile(filePath, structs, document.getText());
+    // In rare cases, the following struct may be undefined. Typically in tests
+    // where roots are changed rapidly.
+    let struct =  structs[filePath];
+    struct = hierachyStructFormat(struct);
+    return struct;
+}
+
+// Create Tree-like structure from flat TexElement[]
+function hierachyStructFormat(struct: TeXElement[]): TeXElement[] {
+    const newStruct: TeXElement[] = [];
+    const orderFromDefaultStructure = new Map<string, number>();
+    for (const [i, section] of defaultStructure.entries()) {
+        orderFromDefaultStructure.set(section, i);
+    }
+    let prev: TeXElement | undefined;
+    for (const section of struct) {
+        // Root Node
+        if (prev === undefined) {
+            newStruct.push(section);
+            prev = section;
+            continue;
+        }
+        // Comparing the order of current section and previous section from defaultStructure
+        if ((orderFromDefaultStructure.get(section.name) ?? defaultStructure.length) > (orderFromDefaultStructure.get(prev.name) ?? defaultStructure.length)) {
+            prev.children.push(section);
+        } else {
+            newStruct.push(section);
+            prev.children = hierachyStructFormat(prev.children);
+            prev = section;
+        }
+    }
+    if (prev !== undefined) {
+        prev.children = hierachyStructFormat(prev.children);
+    }
+    return newStruct;
+}
+
+
+// reference: https://github.com/James-Yu/LaTeX-Workshop/blob/master/src/outline/structurelib/latex.ts#L55
+async function constructFile(filePath: string, structs: FileStructureCache, fileContent:string): Promise<void> {
+    if (structs[filePath] !== undefined) {
+        return;
+    }
+    // Get a list of rnw child chunks
+
+    // Parse each base-level node. If the node has contents, that function
+    // will be called recursively.
+    const rootElement = { children: [] };
+    structs[filePath] = rootElement.children;
+    let ast = await unifiedParser.parse(fileContent);
```

Comment:
The code still replies on AST parse instead of some Regex rules or state machine.
I don't like so many direct cite of LaTeX Workshop code in this PR.

I will wait for a more reasonable explanation until further review.